include/common/coredump.h

```diff
@@ -0,0 +1,279 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2019-2022 Second State INC
+
+//===-- wasmedge/common/coredump.h - Executor coredump definition -----===//
+//
+// Part of the WasmEdge Project.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the coredump class of runtime.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "ast/section.h"
+#include "ast/segment.h"
+#include "ast/type.h"
+#include "common/configure.h"
+#include "common/types.h"
+#include "loader/serialize.h"
+#include "runtime/stackmgr.h"
+#include <cstdint>
+#include <iostream>
+#include <spdlog/spdlog.h>
+#include <string_view>
+#include <vector>
+namespace WasmEdge {
+namespace Coredump {
+
+class Coredump {
+public:
+  Coredump() {}
+  ~Coredump() = default;
+
+  void generateCoredump(Runtime::StackManager &StackMgr) {
+    const Configure Conf;
+    Loader::Serializer Ser(Conf);
+    const auto *CurrentInstance = StackMgr.getModule();
+
+    auto Core = collectProcessInformation(CurrentInstance);
+    // auto DataSection = collectDataSection(StackMgr);
+    auto MemSec = collectMemory(StackMgr);
+    // auto Globals = collectGlobals(StackMgr);
+    auto CoreStack = collectCoreStack(StackMgr);
+    // auto CoreInstances = collectCoreInstances(StackMgr);
+    // auto CoreModules = collectCoreModules(StackMgr);
+
+    //  final file generation
+    AST::Module Mod{};
+    std::vector<Byte> &Magic = Mod.getMagic();
+    Magic.insert(Magic.begin(), {0x00, 0x61, 0x73, 0x6d});
+
+    std::vector<Byte> &Version = Mod.getVersion();
+    Version.insert(Version.begin(), {0x01, 0x00, 0x00, 0x00});
+
+    Mod.getCustomSections().emplace_back(Core);
+    // Mod.getCustomSections().emplace_back(CoreModules);
+    // Mod.getCustomSections().emplace_back(CoreInstances);
+    Mod.getCustomSections().emplace_back(CoreStack);
+    // Mod.getDataSection() = std::move(DataSection);
+    Mod.getMemorySection() = std::move(MemSec);
+    // TODO globals works with DWARF data, is this section needed?
+    //  Mod.getGlobalSection() = std::move(Globals);
+    auto Res = Ser.serializeModule(Mod);
+    std::ofstream File("coredump.wasm", std::ios::out | std::ios::binary);
+    if (File.is_open()) {
+      File.write(reinterpret_cast<const char *>(Res->data()), Res->size());
+      File.close();
+    } else {
+      // Handle error opening file
+      throw std::ios_base::failure("Failed");
```

Comment:
You should not use `throw` without any try-catch block. Instead, use `excepted<T>` for the return type, allowing you to handle the error.